File: java/acorn-core/src/main/java/com/ardley/acorn/permission/ScopeFilter.java
package com.ardley.acorn.permission;

import com.ardley.acorn.attribute.AttributeSource;
import com.fasterxml.jackson.databind.JsonNode;

import java.util.HashMap;
import java.util.Map;

/**
 * A multi-dimensional scope filter that restricts an action to resources matching
 * specific attribute conditions.
 *
 * <p>All dimensions are evaluated with AND semantics — every filter in the map must
 * match for the scope to be satisfied. An empty filter matches everything.
 *
 * <p>JSON representation:
 * <pre>{@code
 * {
 *   "department": {"match": "principal"},
 *   "status": {"equals": "active"},
 *   "region": {"in": ["US", "EU"]}
 * }
 * }</pre>
 */
public record ScopeFilter(Map<String, AttributeFilter> filters) {

    /**
     * Evaluates whether the resource satisfies all filter dimensions given the principal.
     *
     * @param principal the requesting principal's attributes
     * @param resource the target resource's attributes
     * @return true if all dimensions match
     */
    public boolean matches(AttributeSource principal, AttributeSource resource) {
        if (filters.isEmpty()) {
            return true;
        }
        return filters.entrySet().stream()
                .allMatch(entry -> entry.getValue().matches(entry.getKey(), principal, resource));
    }

    /**
     * Parses a scope filter from a JSON object where each key is a dimension name.
     */
    public static ScopeFilter fromJson(JsonNode node) {
        Map<String, AttributeFilter> filters = new HashMap<>();
        var fields = node.fields();
        while (fields.hasNext()) {
            var entry = fields.next();
            filters.put(entry.getKey(), AttributeFilter.fromJson(entry.getValue()));
        }
        return new ScopeFilter(Map.copyOf(filters));
    }
}

File: node/packages/core/src/scope-filter.ts
import type { AttributeSource } from './types.js';
import type { AttributeFilter } from './attribute-filter.js';
import { matchesFilter, parseAttributeFilter } from './attribute-filter.js';

export class ScopeFilter {
  constructor(readonly filters: ReadonlyMap<string, AttributeFilter>) {}

  matches(principal: AttributeSource, resource: AttributeSource): boolean {
    if (this.filters.size === 0) return true;
    for (const [dimension, filter] of this.filters) {
      if (!matchesFilter(filter, dimension, principal, resource)) return false;
    }
    return true;
  }

  static fromObject(obj: unknown): ScopeFilter {
    if (obj === null || typeof obj !== 'object') {
      throw new Error('Scope filter must be an object');
    }
    const filters = new Map<string, AttributeFilter>();
    for (const [key, value] of Object.entries(obj as Record<string, unknown>)) {
      filters.set(key, parseAttributeFilter(value));
    }
    return new ScopeFilter(filters);
  }
}
